Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs

```diff
@@ -44,6 +59,146 @@ public void NormalizeAllPointsThatExceedAllowedLimit()
+                .Select(ps => ps.ProblemsCount)
+                .Max();
+
+            return new CategoryContestsParticipationSummary
+            {
+                MaxProblemsCount = maxProblemsCount,
+                Results = categoryResults,
+            };
+        }
+        
+        public ParticipationsSummaryServiceModel GetParticipationSummaryForContest(Contest contest, bool official)
+        {
+            var participants = 
+                        participantsData.GetAllByContestAndIsOfficial(contest.Id, official)
+                            .Include(p => p.User)
+                            .Include(p => p.Submissions)
+                            .Include(p => p.Scores)
+                            .Include(p => p.Problems)
+                    .ToList();
+            
+            var results = participants
+                .Where(p => p.Scores.Count != 0 && p.Scores.All(ps => ps.Submission != null))
+                .Select(this.SummarizeParticipationWithMaximumPoints)
+                .ToList();
+
+            var problemsCount = contest.ProblemGroups.Count(pg => !pg.Problems.IsNullOrEmpty());
+            
+            return new ParticipationsSummaryServiceModel
+            {
+                ContestId = contest.Id,
+                ContestName = contest.Name,
+                Results = results,
+                ProblemsCount = problemsCount
+            };
+        }
+
+        private ParticipantScoresSummaryModel SummarizeParticipationWithMaximumPoints(Participant participant)
+        {
+            Dictionary<int, DateTime> problemOrderToFirstBestSubmissionTimeCreated = this.CalculateFirstBestSubmissionTimeForProblems(participant);
+
+            var userStartTime = participant.ParticipationStartTime ?? participant.CreatedOn;
+
+            SortedDictionary<int, double> problemOrderToTimeTakenBetweenBest =
+                this.CalculateTimeTakenBetweenBestForProblems(problemOrderToFirstBestSubmissionTimeCreated,
+                    userStartTime);
+
+            var lastSubmission = this.submissionsData.GetLastSubmittedForParticipant(participant.Id);
+
+            return new ParticipantScoresSummaryModel
+            {
+                ParticipantName = participant.User.UserName,
+                ProblemOrderToMinutesTakenToSolve = problemOrderToTimeTakenBetweenBest,
+                PointsTotal = participant.Scores.Select(s => s.Points).Sum(),
+                TimeTotal = Math.Round((lastSubmission.CreatedOn - userStartTime).TotalMinutes),
+            };
+        }
+
+        private Dictionary<int, DateTime> CalculateFirstBestSubmissionTimeForProblems(Participant participant)
+        {
+            Dictionary<int, DateTime> problemOrderToFirstBestSubmissionTimeCreated = new Dictionary<int, DateTime>();
+
+            var participationProblemsForUser = participant.Scores
+                .Select(ps => ps.Problem)
+                .OrderBy(p => p.OrderBy);
+
+            int problemIndex = 0;
```

Comment:
No need for this, since `ForEach` has an overload that passed `index` as parameter

Reply:
not in this version of .net